### PR TITLE
Users Index

### PR DESCRIPTION
**Pull request by redwarewolf**
## Summary

- Added users index
- Added users index tests

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -19,3 +19,9 @@ exports.sessions = (req, res, next) =>
       res.send({ user_id: authorization.user_id });
     })
     .catch(error => next(error));
+
+exports.index = (req, res, next) =>
+  usersService
```
Comment (picollomartin) [inline app/controllers/users.js:24]: Improve naming, this is more like a getUsers endpoint

File: app/services/users.js [hunk 2]
```diff
@@ -28,3 +29,20 @@ exports.createSession = async params => {
   logger.error('Password and mail mismatch for user:', params.mail);
   throw errors.authenticationError("The password and mail combination doesn't match");
 };
+
+exports.index = async req => {
+  logger.info('Index Users with: ', req);
```
Comment (picollomartin) [inline app/services/users.js:34]: This is useful ?

File: app/services/users.js [hunk 3]
```diff
@@ -28,3 +29,20 @@ exports.createSession = async params => {
...
+  const page = req.headers.page || 0;
+  const type =
+    (await jwt.findUserByToken(req.headers.authorization)).type === 'regular'
```
Comment (picollomartin) [inline app/services/users.js:37]: Each time you want the user type you have to do this ? (A query in the database and decode the previous decoded token)

File: app/services/users.js [hunk 4]
```diff
@@ -28,3 +29,20 @@ exports.createSession = async params => {
...
+      ? 'regular'
+      : ['admin', 'regular'];
+  const { count, rows } = await User.findAndCountAll({
+    where: {
```
Comment (picollomartin) [inline app/services/users.js:41]: Be careful with the database errors, seems more proper for a service

File: app/services/users.js [hunk 5]
```diff
@@ -28,3 +29,20 @@ exports.createSession = async params => {
...
+      type
+    },
+    offset: page,
+    limit: config.common.api.paginationLimit
```
Comment (picollomartin) [inline app/services/users.js:45]: Why I can't set my limit (it's ok have a upper limit) but maybe I want less records

File: test/controllers/user.test.js [hunk 6]
```diff
@@ -150,4 +150,67 @@ describe('Users Controller', () => {
...
+    describe('when using an regular user authentication', () => {
+      it('should list only regular users', done => {
+        factory.createMany('User', 5, { type: 'admin' }).then(() => {
+          factory.create('User', { mail: 'pedro.jara@wolox.com.ar', type: 'regular' }).then(() => {
+            request
+              .get('/api/v1/users')
+              .set({
+                Accept: 'application/json',
+                authorization:
+                  'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJtYWlsIjoicGVkcm8uamFyYUB3b2xveC5jb20uYXIifQ.zLLy2i25xQZuXyk0s98afCQA4hlomRq92D1lZQcP-mE'
```
Comment (picollomartin) [inline test/controllers/user.test.js:185]: You have repeated boilerplate between tests (authorization, factories, etc) create some helpers functions

File: Procfile [hunk 7]
```diff
@@ -1 +1 @@
-web: node server.js
+web: npm run start
```
Comment (picollomartin) [inline Procfile:1]: Be careful npm start runs with nodemon, you should keep the original one

File: app/services/users.js (hunk 2 above)
Comment (redwarewolf) [reply app/services/users.js:34]: Old habit from Nubi :P

File: app/middlewares/validators/adminValidator.js [hunk 8]
```diff
@@ -1,15 +1,12 @@
...
 exports.validateAdmin = (req, res, next) => {
   try {
-    userService.findUserByMail(jwt.validate(req.headers.authorization).mail).then(user => {
-      if (user.type !== 'admin') {
-        next(errors.authenticationError('User is not an Admin'));
-      }
-      next();
-    });
+    if (jwt.validate(req.headers.authorization).type !== 'admin') {
+      next(errors.authenticationError('User is not an Admin'));
```
Comment (picollomartin) [inline app/middlewares/validators/adminValidator.js:7]: Add a return here, because the flow goes on

File: app/services/users.js [hunk 9]
```diff
@@ -23,10 +24,32 @@ exports.createSession = async params => {
...
   logger.error('Password and mail mismatch for user:', params.mail);
   throw errors.authenticationError("The password and mail combination doesn't match");
 };
 
+exports.getUsers = async req => {
+  try {
+    const page = req.headers.page || 0;
+    const limit = req.headers.limit || config.common.api.paginationLimit;
+    const type =
+      jwt.validate(req.headers.authorization).type === 'regular' ? 'regular' : ['admin', 'regular'];
```
Comment (picollomartin) [inline app/services/users.js:41]: Add the user info to the request (type, id, etc)

File: app/services/users.js [hunk 10]
```diff
@@ -23,10 +24,32 @@ exports.createSession = async params => {
...
+    };
   }
   logger.error('Password and mail mismatch for user:', params.mail);
   throw errors.authenticationError("The password and mail combination doesn't match");
 };
 
+exports.getUsers = async req => {
+  try {
+    const page = req.headers.page || 0;
+    const limit = req.headers.limit || config.common.api.paginationLimit;
```
Comment (picollomartin) [inline app/services/users.js:39]: Request is processed only in controllers not in services

File: test/controllers/user.test.js [hunk 11]
```diff
@@ -150,4 +151,65 @@ describe('Users Controller', () => {
...
+                done();
+              });
+          });
+        });
+      });
+    });
+
+    describe('when authentication header is missing', () => {
+      it('should not list users', done => {
+        factory.createMany('User', 5, { type: 'admin' }).then(() => {
```
Comment (picollomartin) [inline test/controllers/user.test.js:199]: You could use a before each for this

**Pull request merged**